packages/dev/core/src/ShadersWGSL/ShadersInclude/lightUboDeclaration.fx

```diff
@@ -0,0 +1,98 @@
+#ifdef LIGHT{X}
+	struct Light{X}
+	{
+		vLightData: vec4f,
+		vLightDiffuse: vec4f,
+		vLightSpecular: vec4f,
+		#ifdef SPOTLIGHT{X}
+			vLightDirection: vec4f,
+			vLightFalloff: vec4f,
+		#elif defined(POINTLIGHT{X})
+			vLightFalloff: vec4f,
+		#elif defined(HEMILIGHT{X})
+			vLightGround: vec3f,
+		#endif
+		shadowsInfo: vec4f,
+		depthValues: vec2f
+	} ;
+
+var<uniform> light{X} : Light{X};
+
+
+#ifdef PROJECTEDLIGHTTEXTURE{X}
+	uniform textureProjectionMatrix{X}: mat4x4f;
+	var projectionLightSampler{X}: sampler;
+#endif
+#ifdef SHADOW{X}
+	#ifdef SHADOWCSM{X}
+		uniform lightMatrix{X}:  array<mat4x4f, SHADOWCSMNUM_CASCADES{X}>;
+		uniform viewFrustumZ{X}:  array<f32, SHADOWCSMNUM_CASCADES{X}>;
+        uniform frustumLengths{X}:  array<f32, SHADOWCSMNUM_CASCADES{X}>;
+        uniform cascadeBlendFactor{X}: f32;
+
+		// Because WGSL does not allow us to have arrays as varying...
+		varying vPositionFromLight{X}_0: vec4f;
+		varying vDepthMetric{X}_0:  f32;
+		varying vPositionFromLight{X}_1: vec4f;
+		varying vDepthMetric{X}_1:  f32;
+		varying vPositionFromLight{X}_2: vec4f;
+		varying vDepthMetric{X}_2:  f32;
+		varying vPositionFromLight{X}_3: vec4f;
+		varying vDepthMetric{X}_3:  f32;
+		varying vPositionFromCamera{X}: vec4f;
+
+		var<private> vPositionFromLight{X}: array<vec4f, 4>;
+		var<private> vDepthMetric{X} : array<f32, 4>;
+
+		#if defined(SHADOWPCSS{X})
+			var shadow{X}Sampler: sampler_comparison;			
+			var shadow{X}Texture: texture_depth_2d_array;
+			var depth{X}Sampler: sampler;
+			var depth{X}Texture: texture_2d_array<f32>;

```

Comment:
Given my comments above in `materialHelper.functions.ts`, this should be:
```glsl
var shadowSampler{X}Sampler: sampler_comparison;			
var shadowSampler{X}: texture_depth_2d_array;
var depthSampler{X}Sampler: sampler;
var depthSampler{X}: texture_2d_array<f32>;
```
Same changes to perform below for the `SHADOWPCF` case. This way, the sampler will be automatically bound by the system when the texture is bound.

You could also change `shadowSampler{X}` to `shadowTexture{X}` for the texture name, but the sampler would have to be named `shadowTexture{X}Sampler`. If you make this change, you will also have to do it in the GLSL `lightUboDeclaration.fx` file, as well as update `PrepareUniformsAndSamplersForLight`.

Reply:
Yeah big changes but worth it!